### PR TITLE
Creating dropwdown for selecting category

### PR DESCRIPTION
**Pull request by Icon1cc**

### CODE REVIEW

**Comment by DoStini** [comment]
The dropdown style itself should be the same as the complexity one. Check the "styling" of that one and recreate it

image

**Review by DoStini — CHANGES_REQUESTED** [review]

**Comment by Icon1cc** [comment]
> Please fix the build problem (in the github checks). You can check locally using npm build or npm run build

Fixed this and the build is working now after passing the tests.

File: src/components/CreateExercise.tsx [hunk 1]
```diff
@@ -1,20 +1,47 @@
...
   const navigate = useNavigate();
 
-
+  useEffect(() => {
+    // Set default values for the parameters
+    const defaultPage = 1;
+    const defaultPerPage = 10;
+    const defaultSearchQuery = '';
+    const defaultSortField = '';
+    const defaultSortDir = 'asc'; // or 'desc' or '', depending on your API
```
[the remark this answers is not in the document]
Comment (Icon1cc) [reply src/components/CreateExercise.tsx]: This is fixed now.

File: src/components/CreateExercise.tsx [hunk 2]
```diff
@@ -1,20 +1,47 @@
...
+        const categoryOptions: CategoryOption[] = response.data.map((item: any) => ({
+          value: item.category,
+          label: item.category,
+        }));
+        setCategories(categoryOptions);
+      })
+      .catch((error) => {
+        console.error("Failed to fetch categories", error);
+      });
+  }, []);
```
[the remark this answers is not in the document]
Comment (Icon1cc) [reply src/components/CreateExercise.tsx]: Updated as per your review

File: src/components/ExerciseForm.tsx [hunk 3]
```diff
@@ -1,73 +1,107 @@
...
         </div>
-        <div className="flex-1 p-4  h-1/2 w-3/4 m-auto">
+        <div className="flex-1 p-4 w-3/4">
+          <div>
+            <label className="block mt-2 mb-3 px-1 text-3xl font-bold text-custom-black">
+              Category
+            </label>
+            <Select
+              options={categories}
+              value={categories.find(c => c.value === selectedCategory)}
```
[the remark this answers is not in the document]
Comment (Icon1cc) [reply src/components/ExerciseForm.tsx]: Fixed it

File: src/components/CreateExercise.tsx [hunk 4]
```diff
@@ -1,38 +1,61 @@
...
-      })
-      .catch((error) => {
-        console.log(error);
-      });
+    try {
+      await exercisesService.create(title, textExercise, complexity);
+      navigate("/exercises");
+    } catch (error) {
+      console.error("Failed to create exercise:", error);
+    }
```
Comment (DoStini) [inline src/components/CreateExercise.tsx:39]: Why this change?

File: src/components/EditExercise.tsx [hunk 5]
```diff
@@ -1,57 +1,72 @@
...
+          label: category,
+        }));
+        setCategories(categoryOptions);
+
+        if (id) {
+          const exerciseResponse = await exerciseService.getByID(id);
+          setTitle(exerciseResponse.data.title);
+          setComplexity(exerciseResponse.data.complexity);
+          setTextExercise(exerciseResponse.data.text);
+          setSelectedCategory(exerciseResponse.data.category); // Assuming the response has a 'category' field
```
Comment (DoStini) [inline src/components/EditExercise.tsx:39]: Why this change?

File: src/components/EditExercise.tsx [hunk 6]
```diff
@@ -1,57 +1,72 @@
...
+  }, [id, navigate]);
 
   const handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
     e.preventDefault();
-    id && exerciseService
-      .edit(id, title, textExercise, complexity)
-      .then(() => {
+    if (id) {
+      try {
+        await exerciseService.edit(id, title, textExercise, complexity);
```
Comment (DoStini) [inline src/components/EditExercise.tsx:56]: Why this change?

**Review by DoStini — CHANGES_REQUESTED** [review]

File: src/components/CreateExercise.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (DoStini) [reply src/components/CreateExercise.tsx]: Not resolved

File: src/components/ExerciseForm.tsx [hunk 7]
```diff
@@ -1,73 +1,127 @@
...
+        <Select
+          id="category"
+          options={categories}
+          value={categories.find(c => c.value === selectedCategory)}
+          onChange={handleCategoryChange}
+          styles={customStyles}
+          className="text-lg"
+          placeholder="Select Category"
+          required
+        />
```
Comment (DoStini) [inline src/components/ExerciseForm.tsx:100]: Export the select to a different component, also the custom styles aswell and the get categories API calls, it's getting way too confusing and big this file

**Review by DoStini — CHANGES_REQUESTED** [review]

File: src/components/CategorySelect.tsx [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+const CategorySelect = ({ categories, selectedCategory, setSelectedCategory }: CategorySelectProps) => {
+  const handleCategoryChange = (selectedOption: CategoryOption | null) => {
+    setSelectedCategory(selectedOption ? selectedOption.value : '');
+  };
+
+  return (
+    <Select
+      options={categories}
+      value={categories.find(c => c.value === selectedCategory)}
```
Comment (DoStini) [inline src/components/CategorySelect.tsx:45]: This is a very costly operation, we should refactor selectedCategory to be not only the value but the object itself
```suggestion
      value={selectedCategory}
```

But here `selectedCategory` must be `{value, label}`

File: src/components/CategorySelect.tsx [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+      onChange={handleCategoryChange}
+      styles={customStyles}
+      className="text-lg"
+      placeholder="Select Category"
```
Comment (DoStini) [inline src/components/CategorySelect.tsx:49]: Repeating what I said, there is this component `  <AsyncSelect cacheOptions loadOptions={loadOptions} defaultOptions />` presented in the docs which is more usefull in this case, since we're dealing with promises.

So I suggest "creating the promise" as you did, but only for the categories, and bring that function inside this component. And then using that function to the "loadOptions"

Refer to the example in the documentation page: https://react-select.com/async

**Review by DoStini — CHANGES_REQUESTED** [review]

File: src/components/ExerciseForm.tsx [hunk 10]
```diff
@@ -1,73 +1,95 @@
...
+          type="text"
+          className="h-10 px-3 mb-4 shadow-lg border border-custom-blue text-custom-black sm:text-sm rounded-lg focus:ring-primary-600 focus:border-primary-600 block w-full"
+          placeholder="Add Title"
+          required
+        />
+      </div>
 
-    <form onSubmit={onSubmit}>
-      <div className="flex flex-col mt-10 items-center justify-center h-3/4 w-3/4 m-auto shadow-lg shadow-custom-grey sm:max-w-[900px] bg-custom-light-grey">
-        <div className="flex h-1/5 w-3/4 mt-0 m-auto ">
```
Comment (DoStini) [inline src/components/ExerciseForm.tsx:21]: Restore this I think it looked better before

**Review by DoStini — CHANGES_REQUESTED** [review]

File: src/components/CreateExercise.tsx [hunk 11]
```diff
@@ -1,38 +1,65 @@
...
+      const categoryOptions: CategoryOption[] = response.data.map((category: string) => ({
+        value: category,
+        label: category,
+      }));
+      return categoryOptions;
+    } catch (error) {
+      console.error("Failed to fetch categories", error);
+      return [];
+    }
+  };
```
Comment (DoStini) [inline src/components/CreateExercise.tsx:32]: Perfect, just move this to the category select component and it's good to go! : )

File: src/components/CreateExercise.tsx [hunk 12]
```diff
@@ -1,38 +1,65 @@
...
-    exerciseService
-      .create(title, textExercise, complexity)
-      .then(() => {
-        navigate("/exercises");
-      })
-      .catch((error) => {
-        console.log(error);
-      });
+    try {
+      if (selectedCategory) {
```
Comment (DoStini) [inline src/components/CreateExercise.tsx:37]: This can be removed if the category dropdown has the prop required

File: src/components/CategorySelect.tsx [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+};
+
+const CategorySelect = ({ loadOptions, selectedCategory, setSelectedCategory }: CategorySelectProps) => {
+  const handleCategoryChange = (selectedOption: CategoryOption | null) => {
+    setSelectedCategory(selectedOption); 
+  };
+
+  return (
+    <AsyncSelect
+      cacheOptions
```
Comment (DoStini) [inline src/components/CategorySelect.tsx:44]: add `required`

**Review by DoStini — CHANGES_REQUESTED** [review]

**Review by DoStini — CHANGES_REQUESTED** [review]

File: src/components/CreateExercise.tsx (hunk 12 above)
Comment (Icon1cc) [reply src/components/CreateExercise.tsx:17]: Done

File: src/components/CreateExercise.tsx [hunk 14]
```diff
@@ -1,40 +1,45 @@
...
+        title={title}
+        setTitle={setTitle}
+        complexity={complexity}
+        setComplexity={setComplexity}
+        textExercise={textExercise}
+        setTestExercise={setTextExercise}
+        selectedCategory={selectedCategory}
+        setSelectedCategory={setSelectedCategory} loadOptions={function (): Promise<{ value: string; label: string; }[]> {
+          throw new Error("Function not implemented.");
+        } } />
```
Comment (DoStini) [inline src/components/CreateExercise.tsx:40]:
```suggestion
        setSelectedCategory={setSelectedCategory}
        />
```

File: src/components/EditExercise.tsx [hunk 15]
```diff
@@ -1,57 +1,64 @@
...
+    const fetchData = async () => {
+      setIsLoading(true);
+      try {
+        if (id) {
+          const exerciseResponse = await exerciseService.getByID(id);
+          setTitle(exerciseResponse.data.title);
+          setComplexity(exerciseResponse.data.complexity);
+          setTextExercise(exerciseResponse.data.text);
+          const category = exerciseResponse.data.category[0]?.category;
+          console.log(category);
```
Comment (DoStini) [inline src/components/EditExercise.tsx:30]:
```suggestion
```

File: src/components/EditExercise.tsx [hunk 16]
```diff
@@ -61,14 +68,18 @@ const EditExercise = () => {
...
         setComplexity={setComplexity}
+        textExercise={textExercise}
         setTestExercise={setTextExercise}
-        title={title}
-        complexity={complexity}
-        textExercise={textExercise} />
+        selectedCategory={selectedCategory}
+        setSelectedCategory={setSelectedCategory} loadOptions={function (): Promise<{ value: string; label: string; }[]> {
+          throw new Error("Function not implemented.");
+        }} />
```
Comment (DoStini) [inline src/components/EditExercise.tsx:80]:
```suggestion
        setSelectedCategory={setSelectedCategory}
        />
```

File: src/components/ExerciseForm.tsx [hunk 17]
```diff
@@ -1,73 +1,89 @@
...
   complexity: string;
   textExercise: string;
-  setTitle: React.Dispatch<React.SetStateAction<string>>
-  setComplexity: React.Dispatch<React.SetStateAction<string>>
-  setTestExercise: React.Dispatch<React.SetStateAction<string>>
-}
-
-const ExerciseForm = ({ onSubmit, title, complexity, textExercise, setTitle, setComplexity, setTestExercise }: ExerciseFormProps) => {
-
+  loadOptions: (inputValue: string) => Promise<CategoryOption[]>;
```
Comment (DoStini) [inline src/components/ExerciseForm.tsx:15]:
```suggestion
```

**Review by DoStini — CHANGES_REQUESTED** [review]

**Review by DoStini — APPROVED** [review]
